lib/tasks/green_lanes.rake

```diff
@@ -191,4 +191,53 @@ namespace :green_lanes do
       end
     end
   end
+
+  desc 'Add CategoryAssessments CATEGORY_ASSESSMENT_CSV_FILE=path/to/file.csv'
```

Comment:
my only worry about long rake tasks like this is its quite a lot going on and whether breaking it down something like this might be more readable:

```
desc 'Add CategoryAssessments CATEGORY_ASSESSMENT_CSV_FILE=path/to/file.csv'
task import_csv_category_assessments: :environment do
  raise 'Only supported on XI service' unless TradeTariffBackend.xi?

  csv_data = fetch_csv_data
  parsed_data = CSV.parse(csv_data, headers: true)

  assessments = GreenLanes::CategoryAssessment.all.index_by do |assessment|
    [assessment.measure_type_id, assessment.regulation_id, assessment.theme_id]
  end

  parsed_data.each do |row|
    process_row(row, assessments)
  end
end

def fetch_csv_data
  if Rails.application.config.persistence_bucket.present?
    fetch_csv_from_s3
  else
    fetch_csv_from_local_file
  end
end

def fetch_csv_from_s3
  object_key = 'data/categorisation/category_assessment.csv'
  Rails.application.config.persistence_bucket.object(object_key).get.body.read
rescue Aws::S3::Errors::NoSuchKey => e
  raise InvalidFile, "File not found in S3 (#{e.message})"
end

def fetch_csv_from_local_file
  file_path = ENV['CATEGORY_ASSESSMENT_CSV_FILE'].to_s
  raise "Cannot read file '#{file_path}'" unless File.file?(file_path)

  File.read(file_path)
end

def process_row(row, assessments)
  theme = GreenLanes::Theme.find_by(section: row['theme_section'], subsection: row['theme_subsection'])
  key = [row['measure_type_id'], row['regulation_id'], theme.id]
  assessment = assessments[key] || initialize_new_assessment(row, theme)

  if assessment.id.nil?
    assessment.theme = theme
    assessment.save(validate: true)
    add_exemption_to_assessment(row['exemption_code'], assessment)
  end

  assessments[key] = assessment
end

def initialize_new_assessment(row, theme)
  GreenLanes::CategoryAssessment.new.tap do |assessment|
    assessment.measure_type_id = row['measure_type_id']
    assessment.regulation_id = row['regulation_id']
    assessment.regulation_role = row['regulation_role']
    assessment.theme = theme
  end
end

def add_exemption_to_assessment(exemption_code, assessment)
  exemption = GreenLanes::Exemption.find_by(code: exemption_code)
  assessment.add_exemption(exemption) if exemption
end

Reply:
Fyi i havent tested the above hence why might be easier to explain on call tomorrow what going on?